File: Python/Contest 3/2019 - 2020/VeitchIntermediate.py
#
# ACSL 2019-2020 - Contest 3 - Veitch - Intermediate Division
#
#

def get_veitch_diagram(boolean_expression):

    veitch_diagram = [0] * 16
    grid = [
        "1100", "1110", "0110", "0100", "1101", "1111", "0111", "0101",
        "1001", "1011", "0011", "0001", "1000", "1010", "0010", "0000"
    ]
    expressions = boolean_expression.replace("+", " ").split(" ")

    for expression in expressions:
        binary_expression = convert_expression(expression)
        for k in range(16):
            if check_match(binary_expression, grid[k]):
                veitch_diagram[k] = 1

    result = ""
    for k in [0, 4, 8, 12]:
        value = 8 * veitch_diagram[k] + 4 * veitch_diagram[k+1] + 2 * veitch_diagram[k+2] + veitch_diagram[k+3]
        if value < 10:
            result += str(value)
        else:
            result += chr(ord('A') + value - 10)

    return result


def convert_expression(expression):

    binary_expression = ""

    for c in "ABCD":
        if ("~" + c) in expression:
            binary_expression += "0"
        elif c in expression:
            binary_expression += "1"
        else:
            binary_expression += "*"

    return binary_expression


def check_match(binary_expression, cell):

    match = True
    for j in range(4):
        if binary_expression[j] != '*' and binary_expression[j] != cell[j]:
            match = False
            break

    return match


test_input = [
    "AB+~AB+~A~B", "AB~C~D+AB~CD+~A~B~CD",
    "AB~C~D+~AB~C~D+A~B~C~D", "B~D+~B~D",
    "~A~BD+~A~B~D", "B~D+~A~BD+A~B~C",
    "~B~C+BCD+B~C~D", "A~C+ACD+~A~CD",
    "AB~D+~ABD+A~BD+~A~B~D", "B~D+~A~CD+~A~B~C~D"
]

test_output = ["FF33", "8810", "9008", "F00F", "0033", "F0B8", "9699", "8DD8", "C3C3", "F111"]

for i in range(10):
    test_result = get_veitch_diagram(test_input[i])

    if test_result == test_output[i]:
        print("Test Case " + str(i) + ": Passed!")
    else:
        print("Test Case " + str(i) + ": Failed!")
        print("Expected output: " + test_output[i])
        print("Your output: " + test_result)

File: Java/Contest 3/2019 - 2020/VeitchIntermediate.java
/**
 *      ACSL 2019-2020 - Contest 3 - Veitch - Intermediate Division
 *
 */

import java.io.*;
import java.lang.*;
import java.util.*;

public class VeitchIntermediate {

    static String getVeitchDiagram (String booleanExpression) {

        int[] veitchDiagram = new int[16];
        String[] grid = {
                "1100", "1110", "0110", "0100", "1101", "1111", "0111", "0101",
                "1001", "1011", "0011", "0001", "1000", "1010", "0010", "0000"};
        String[] expressions = booleanExpression.replace("+", " ").split(" ");

        for (String expression : expressions)
        {
            String binaryExpression = convertExpression(expression);
            for (int i = 0; i < 16; i++)
            {
                if (checkMatch(binaryExpression, grid[i])) {
                    veitchDiagram[i] = 1;
                }
            }
        }

        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < 16; i=i+4)
        {
            int value = 8 * veitchDiagram[i] + 4 * veitchDiagram[i+1] + 2 * veitchDiagram[i+2] + veitchDiagram[i+3];
            if (value < 10) {
                sb.append(value);
            }
            else {
                sb.append((char)('A' + value - 10));
            }
        }

        return sb.toString();
    }

    static String convertExpression (String expression) {

        StringBuilder sb = new StringBuilder();
        String inputs = "ABCD";

        for (int i = 0; i < 4; i++)
        {
            String input = "" + inputs.charAt(i);
            if (expression.contains("~" + input)) {
                sb.append("0");
            }
            else if (expression.contains(input)) {
                sb.append("1");
            }
            else {
                sb.append("*");
            }
        }

        return sb.toString();
    }

    static boolean checkMatch (String binaryExpression, String cell) {

        boolean match = true;
        for (int i = 0; i < 4; i++)
        {
            if (binaryExpression.charAt(i) != '*' && binaryExpression.charAt(i) != cell.charAt(i)) {
                match = false;
                break;
            }
        }

        return match;
    }

    public static void main (String [] args) {

        String[] input = new String[10];
        input[0] = "AB+~AB+~A~B";
        input[1] = "AB~C~D+AB~CD+~A~B~CD";
        input[2] = "AB~C~D+~AB~C~D+A~B~C~D";
        input[3] = "B~D+~B~D";
        input[4] = "~A~BD+~A~B~D";
        input[5] = "B~D+~A~BD+A~B~C";
        input[6] = "~B~C+BCD+B~C~D";
        input[7] = "A~C+ACD+~A~CD";
        input[8] = "AB~D+~ABD+A~BD+~A~B~D";
        input[9] = "B~D+~A~CD+~A~B~C~D";

        String[] output = {"FF33", "8810", "9008", "F00F", "0033", "F0B8", "9699", "8DD8", "C3C3", "F111"};

        for (int i = 0; i < 10; i++)
        {
            String result = getVeitchDiagram(input[i]);

            if (output[i].equals(result)) {
                System.out.println("Test Case " + i + ": Passed!");
            }
            else {
                System.out.println("Test Case " + i + ": Failed!");
                System.out.println("Expected output: " + output[i]);
                System.out.println("Your output: " + result);
            }
        }
    }
}